lib/eth/abi/event.rb

```diff
@@ -24,20 +24,69 @@ module Abi
     module Event
       extend self
 
+      # Compute topic for ABI event interface.
+      #
+      # @param interface [Hash] ABI event interface.
+      # @return [String] a hex-string topic.
+      def compute_topic(interface)
+        sig = Abi.signature(interface)
+        Util.prefix_hex(Util.bin_to_hex(Util.keccak256(sig)))
+      end
+
+      # Decodes a stream of receipt logs with a set of ABI interfaces.
+      #
+      # @param interfaces [Array] event ABI types.
+      # @param logs [Array] transaction receipt logs
+      # @return [Hash] an enumerator of decoded log objects.
+      def decode_logs(interfaces, logs)
+        Enumerator.new do |y|
+          topic_to_interfaces = Hash[interfaces.map { |i| [compute_topic(i), i] }]
+
+          logs.each do |log|
+            result = { log: log }
+            topics = log.fetch("topics", [])
+            topic = topics[0]
+
+            if interface = topic_to_interfaces[topic]
+              result[:name] = interface.fetch("name")
+              result[:signature] = Abi.signature(interface)
+              result[:topic] = topic
+
+              inputs = interface.fetch("inputs")
+              data = log.fetch("data")
+              anonymous = interface.fetch("anonymous", false)
+
+              args, kwargs = decode_log(inputs, data, topics, anonymous)
+              result[:args] = args
+              result[:kwargs] = kwargs
+            end
+
+            y << result
+          end
+        end
+      end
+
       # Decodes event log argument values.
       #
       # @param inputs [Array] event ABI types.
       # @param data [String] ABI event data to be decoded.
       # @param topics [Array] ABI event topics to be decoded.
+      # @param anonymous [Boolean] If event signature is excluded from topics.
```

Comment:
Not super useful to me, but added this for completeness. Events can be marked as `"anonymous"` in the ABI and don't put their event signature as the first topic. Haven't seen it used much, but it's [something Solidity supports](https://docs.soliditylang.org/en/v0.8.13/contracts.html?highlight=anonymous#events).